### PR TITLE
Python: Add modeling of `jmespath`

### PR DESCRIPTION
**Pull request by RasmusWL**

### CODE REVIEW

File: python/ql/test/library-tests/frameworks/jmespath/taint_test.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import jmespath
+
+def test_taint():
+    data = TAINTED_DICT
+
+    expression = jmespath.compile("foo.bar")
```
Comment (yoff) [inline python/ql/test/library-tests/frameworks/jmespath/taint_test.py:6]: Calling these `unsafe_data` and `safe_expression` respectively would strengthen the symmetry and make the point clear (that only the data's safety is important).

File: python/ql/src/semmle/python/frameworks/Jmespath.qll [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * Provides models for the `jmespath` PyPI package.
+ * See https://pypi.org/project/jmespath/.
+ */
+private module Jmespath {
+  class JmespathAdditionalTaintSteps extends TaintTracking::AdditionalTaintStep {
+    override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+      exists(DataFlow::CallCfgNode call |
+        call = API::moduleImport("jmespath").getMember("search").getACall() and
+        nodeFrom in [call.getArg(1), call.getArgByName("data")] and
```
Comment (yoff) [inline python/ql/src/semmle/python/frameworks/Jmespath.qll:21]: Out of interest, how did you determine `data` (and `value` below), I could not see them in the documentation.

**Review by yoff — CHANGES_REQUESTED** [review]
Minor naming request, LGTM

File: python/ql/src/semmle/python/frameworks/Jmespath.qll (hunk 2 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Jmespath.qll:21]: I tried it out with locally installed package :wink:

Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Jmespath.qll:21]: and here is the code implementing it:

- https://github.com/jmespath/jmespath.py/blob/develop/jmespath/__init__.py
- https://github.com/jmespath/jmespath.py/blob/599b0f7ea650a60f7512233301e70ff8151de0f0/jmespath/parser.py#L507

File: python/ql/test/library-tests/frameworks/jmespath/taint_test.py (hunk 1 above)
Comment (RasmusWL) [reply python/ql/test/library-tests/frameworks/jmespath/taint_test.py]: I don't really agree that this is too much of a problem, but it was easy to fix, so here you are :blush:

File: python/ql/src/semmle/python/frameworks/Jmespath.qll (hunk 2 above)
Comment (yoff) [reply python/ql/src/semmle/python/frameworks/Jmespath.qll:21]: I was contemplating installing the package..going to the source is what I should have done, thanks.

**Review by yoff — APPROVED** [review]
LGTM

File: python/ql/test/library-tests/frameworks/jmespath/taint_test.py (hunk 1 above)
Comment (yoff) [reply python/ql/test/library-tests/frameworks/jmespath/taint_test.py]: I would not say it was a problem either, rather an opportunity :-) Thanks for the update.

File: python/ql/src/semmle/python/frameworks/Jmespath.qll [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        nodeTo = call
+        or
+        call =
+          API::moduleImport("jmespath")
+              .getMember("compile")
+              .getReturn()
+              .getMember("search")
+              .getACall() and
+        nodeFrom in [call.getArg(0), call.getArgByName("value")] and
+        nodeTo = call
```
Comment (tausbn) [inline python/ql/src/semmle/python/frameworks/Jmespath.qll:31]: You could get rid of the duplication of line 22 by moving `nodeTo = call` into a separate part of the `exists`.

Also, I wonder if you could just change the type of `nodeTo` to `CallCfgNode` and forego the `exists` altogether. (I know overriding predicates can specify a more specific result type, but off the top of my head I don't know if this applies to arguments as well).

Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Jmespath.qll:31]: You're not allowed to provide more specific types for the arguments when overriding.

We could do `exists(DataFlow::CallCfgNode | nodeTo = Call | ...)`, but I feel that the fact that the nodeTo is the _result_ of the call, and not the post-update node of one of the arguments is more accidental than _the way it always must be_ (although it probably will be for +90% of all additional taint steps we write involving calls). I think I'm think I'm just going to leave it as is, but probably wouldn't be too resistant to changing it in the future :wink:

**Comment by RasmusWL** [comment]
:shipit:

**Pull request merged**